beacon-chain/db/kv/migration_state_validators.go

```diff
@@ -0,0 +1,91 @@
+package kv
+
+import (
+	"bytes"
+	"context"
+
+	"github.com/golang/snappy"
+	v1 "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
+	"github.com/prysmaticlabs/prysm/shared/featureconfig"
+	"github.com/prysmaticlabs/prysm/shared/hashutil"
+	bolt "go.etcd.io/bbolt"
+)
+
+var migrationStateValidators0Key = []byte("migration_state_validator")
+
+func migrateStateValidators(tx *bolt.Tx) error {
+	if !featureconfig.Get().EnableHistoricalSpaceRepresentation {
+		return nil
+	}
+	mb := tx.Bucket(migrationsBucket)
+	if b := mb.Get(migrationStateValidators0Key); bytes.Equal(b, migrationCompleted) {
+		return nil // Migration already completed.
+	}
+
+	// get the source and destination buckets.
+	log.Infof("migrating %s. It will take few minutes. Please wait.", stateBucket)
+	stateBkt := tx.Bucket(stateBucket)
+	if stateBkt == nil {
+		return nil
+	}
+	valBkt := tx.Bucket(stateValidatorsBucket)
+	if valBkt == nil {
+		return nil
+	}
+	indexBkt := tx.Bucket(blockRootValidatorHashesBucket)
+	if indexBkt == nil {
+		return nil
+	}
+
+	// for each of the state in the stateBucket, do the migration.
+	ctx := context.Background()
+	c := stateBkt.Cursor()
+	for k, v := c.First(); k != nil; k, v = c.Next() {
+		state := &v1.BeaconState{}
+		if decodeErr := decode(ctx, v, state); decodeErr != nil {
+			return decodeErr
+		}
+
+		// move all the validators in this state registry out to a new bucket.
+		var validatorKeys []byte
+		for _, val := range state.Validators {
+			valBytes, encodeErr := encode(ctx, val)
+			if encodeErr != nil {
+				return encodeErr
+			}
+
+			// create the unique hash for that validator entry.
+			hash := hashutil.Hash(valBytes)
+
+			// add the validator in the stateValidatorsBucket, if it is not present.
+			if valEntry := valBkt.Get(hash[:]); valEntry == nil {
```

Comment:
.Has() is perhaps more efficient? Also, doesn't put handle this logic of adding if not present? Worth a check

Reply:
bolt DB doesn't seem to have Has(). 
More context here https://github.com/boltdb/bolt/issues/321#issuecomment-76822462

As you said, Put handles the "overwrite if present" logic. but since a single validator entry will be written several hundreds of times, checking before write will probably shave off some time.